storage/fanout.go

```diff
@@ -240,23 +240,46 @@ func (q *mergeQuerier) Select(params *SelectParams, matchers ...*labels.Matcher)
 func (q *mergeQuerier) SelectSorted(params *SelectParams, matchers ...*labels.Matcher) (SeriesSet, Warnings, error) {
 	seriesSets := make([]SeriesSet, 0, len(q.queriers))
 	var warnings Warnings
+
+	var priErr error = nil
+	type queryResult struct {
+		qr          Querier
+		set         SeriesSet
+		wrn         Warnings
+		selectError error
+	}
+	queryResultChan := make(chan *queryResult)
 	for _, querier := range q.queriers {
-		set, wrn, err := querier.SelectSorted(params, matchers...)
-		q.setQuerierMap[set] = querier
-		if wrn != nil {
-			warnings = append(warnings, wrn...)
+		go func(qr Querier) {
+			if priErr != nil {
```

Comment:
You've still a data race here, you can remove this if entirely.

Reply:
Thank you for your review.
done.
![image](https://user-images.githubusercontent.com/9583245/74215148-0d207080-4cdb-11ea-8573-8622951b4e29.png)